### PR TITLE
MAINT: stats.logistic.fit: ensure scale is positive

### LINKED ISSUE #18176: BUG: scipy.stats.logistic.fit() returns negative `scale`
### Describe your issue.

Fitting a logistic distribution to data using `scipy.stats.logistic.fit()` returns a negative value for the scale parameter for some input data. Changing the sign of the result and using it as the initial guess for the next `fit()` returns practically the same result as the input values.

The issue happens on versions 1.9.1 and 1.10.1, but based on blame the current head of development has exactly the same code as 1.10.1 does, https://github.com/scipy/scipy/blame/v1.10.1/scipy/stats/_continuous_distns.py#L5341-L5448

It seems that the analytic solution for location and scale parameters involve an even function with respect to scale: https://github.com/scipy/scipy/blob/main/scipy/stats/_continuous_distns.py#L5725-L5727 - although I have not quite worked out how the sign flip does not mess up the `loc` parameter equation.

Maybe the solution should be constrained to strictly positive values of `scale`? We worked around the issue by rerunning `fit()` using `(loc, -scale)` as the initial guess if the first `fit()` returned `scale < 0`, but this might not be ideal. It seems that even in general, the analytical formulas (especially indirect ones relying on finding roots) should explicitly constrain the output values to valid ranges in https://github.com/scipy/scipy/issues/11782 ?

### Reproducing Code Example

```python
import numpy as np
import scipy

data = np.array([-459, 37, 43, 45, 45, 48, 54, 55, 58] + [59] * 3 + [61] * 9)
params = scipy.stats.logistic.fit(data)
print(f"SciPy version: {scipy.__version__}")
print(f"Params (loc, scale): {params}")
```

### Error message

```shell
(none)
```

### SciPy/NumPy/Python version and system information

```shell
1.10.1 1.23.5 sys.version_info(major=3, minor=10, micro=9, releaselevel='final', serial=0)
lapack_mkl_info:
  NOT AVAILABLE
openblas_lapack_info:
  NOT AVAILABLE
openblas_clapack_info:
  NOT AVAILABLE
flame_info:
  NOT AVAILABLE
accelerate_info:
  NOT AVAILABLE
atlas_3_10_threads_info:
  NOT AVAILABLE
atlas_3_10_info:
  NOT AVAILABLE
atlas_threads_info:
  NOT AVAILABLE
atlas_info:
  NOT AVAILABLE
lapack_info:
    libraries = ['lapack', 'blas', 'lapack', 'blas']
    library_dirs = ['/home/[REDACTED]/miniconda3/envs/[ENV_REDACTED]/lib']
    language = f77
blas_mkl_info:
  NOT AVAILABLE
blis_info:
  NOT AVAILABLE
openblas_info:
  NOT AVAILABLE
atlas_3_10_blas_threads_info:
  NOT AVAILABLE
atlas_3_10_blas_info:
  NOT AVAILABLE
atlas_blas_threads_info:
  NOT AVAILABLE
atlas_blas_info:
  NOT AVAILABLE
blas_info:
    libraries = ['cblas', 'blas', 'cblas', 'blas']
    library_dirs = ['/home/[REDACTED]/miniconda3/envs/[ENV_REDACTED]/lib']
    include_dirs = ['/home/[REDACTED]/miniconda3/envs/[ENV_REDACTED]/include']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
blas_opt_info:
    define_macros = [('NO_ATLAS_INFO', 1), ('HAVE_CBLAS', None)]
    libraries = ['cblas', 'blas', 'cblas', 'blas']
    library_dirs = ['/home/[REDACTED]/miniconda3/envs/[ENV_REDACTED]/lib']
    include_dirs = ['/home/[REDACTED]/miniconda3/envs/[ENV_REDACTED]/include']
    language = c
lapack_opt_info:
    libraries = ['lapack', 'blas', 'lapack', 'blas', 'cblas', 'blas', 'cblas', 'blas']
    library_dirs = ['/home/[REDACTED]/miniconda3/envs/[ENV_REDACTED]/lib']
    language = c
    define_macros = [('NO_ATLAS_INFO', 1), ('HAVE_CBLAS', None)]
    include_dirs = ['/home/[REDACTED]/miniconda3/envs/[ENV_REDACTED]/include']
Supported SIMD extensions in this NumPy install:
    baseline = SSE,SSE2,SSE3
    found = SSSE3,SSE41,POPCNT,SSE42,AVX,F16C,FMA3,AVX2,AVX512F,AVX512CD,AVX512_SKX,AVX512_CLX,AVX512_CNL,AVX512_ICL
    not found = AVX512_KNL,AVX512_KNM
```

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Closes gh-18176

#### What does this implement/fix?
gh-18176 reported a case in which the reported solution had `scale < 0`. To fix the bug, ~~we'll use~~ we return `abs(scale)`.
* `dl_dscale` is an even function
* `dl_dloc` ~~appears to be approximately even at the solution~~ is odd, and 
* in 10k randomly-generated tests, the fit override always outperformed the generic fit. This appears to be true even when the data is not distributed according to the logistic distribution and when scale/loc are fixed at random values.

#### Additional information
~~As a future enhancement, we would *constrain* `scale` to be positive. The fixed scale/fixed loc cases can use a bracketing root finder, for instance.~~

<details>

The usual experiment:
```python3
import numpy as np
from scipy import stats
from time import perf_counter
import matplotlib.pyplot as plt
from itertools import product

rng = np.random.default_rng(24330062354926113)
qrng = stats.qmc.Halton(6, seed=rng)
dist = stats.logistic

N = 10000

megadict = {}
i_stop = {}

for fix_params in product((False, True), repeat=2):
    if np.all(fix_params):
        continue

    print(fix_params)

    times_pr = np.full(N, np.nan)
    times_super = np.full(N, np.nan)
    nllfs_super = np.full(N, np.nan)
    nllfs_pr = np.full(N, np.nan)

    for i in range(N):
        fparams = {}

        if i % 50 == 0:
            print(i)

        sample = qrng.random()
        sample = stats.qmc.scale(sample,
                                 [-3, -3, -3, -3, 1, -1],
                                 [+3, +3, +3, +3, 5, +1])[0]
        b, c, loc, scale, size = 10 ** sample[:-1]  # some aren't used here
        c = c + 1
        loc_sign = np.sign(sample[-1])
        loc *= loc_sign
        size = int(size)

        params_true = loc, scale
        rvs = dist.rvs(*params_true, size=size, random_state=rng)

        if fix_params[-2]:
            fparams['floc'] = loc
        if fix_params[-1]:
            fparams['fscale'] = scale

        if i_stop and i not in i_stop:
            continue

        try:
            tic = perf_counter()
            params_fit = dist.fit(rvs, **fparams)
            toc = perf_counter()
            nllf_pr = dist.nnlf(params_fit, rvs)
            nllfs_pr[i] = nllf_pr
            times_pr[i] = toc - tic

        except Exception as e:
            print(f"Failure in {i}: {e}")
            print(*params_fit, size)

        try:
            tic = perf_counter()
            params_fit = dist.fit(rvs, **fparams, superfit=True)
            toc = perf_counter()
            nllf_super = dist.nnlf(params_fit, rvs)
            nllfs_super[i] = nllf_super
            times_super[i] = toc - tic

        except Exception as e:
            print(f"Failure in {i}: {e}")
            print(*params_fit, size)

    megadict[fix_params] = {'times_pr': times_pr,
                            'times_super': times_super,
                            'nllfs_pr': nllfs_pr,
                            'nllfs_super': nllfs_super}

    # print(s, loc, scale, size, nllf_super < nllf_pr)
    # if i in i_stop:
    # break

np.savez("logistic_fitresults", megadict=megadict)


import numpy as np
from numpy.testing import assert_allclose, assert_equal
from scipy import stats, optimize, integrate
import matplotlib.pyplot as plt
from mpmath import mp
mp.dps = 50
rng = np.random.default_rng(1638083107694713882823079058616272161)

megadict = np.load("logistic_fitresults.npz", allow_pickle=True)['megadict'][()]

fig, axs = plt.subplots(1, 3, figsize=(10, 5), sharex=True, sharey=True)
plt.subplots_adjust(hspace=0.25)

for i, item in enumerate(megadict.items()):
    ax = axs.ravel()[i]
    key, val = item

    labels = np.array(['floc', 'fscale'])
    title = ", ".join(labels[np.array(key)]) or "all free"

    nllfs_pr = val['nllfs_pr']
    nllfs_super = val['nllfs_super']
    times_pr = val['times_pr']
    times_super = val['times_super']
    nllf_diff = (nllfs_pr - nllfs_super) / np.abs(nllfs_super)
    nllf_diff = nllf_diff[np.isfinite(nllf_diff)]
    nllf_diff_good = -nllf_diff[nllf_diff < 0]
    nllf_diff_bad = nllf_diff[nllf_diff > 0]
    new_nans = np.isnan(nllfs_pr) & (~np.isnan(nllfs_super))
    title = title + f", {np.sum(new_nans)} new NaNs"

    bins = np.arange(-4.5, 0, 0.25)
    ax.hist(np.log10(times_pr), bins=bins, alpha=0.5)
    ax.hist(np.log10(times_super), bins=bins, alpha=0.5)
    ax.set_title(title)

plt.xlabel('log10 of `fit` execution time (s)')
plt.ylabel('Frequency')
plt.legend(('PR', 'main'))
fig.suptitle('Histogram of log10 of `fit` execution time (s)')
plt.show()

fig, axs = plt.subplots(1, 3, figsize=(10, 5), sharex=True, sharey=True)
plt.subplots_adjust(hspace=0.25)

for i, item in enumerate(megadict.items()):
    ax = axs.ravel()[i]
    key, val = item

    labels = np.array(['floc', 'fscale'])
    title = ", ".join(labels[np.array(key)]) or "all free"

    nllfs_pr = val['nllfs_pr']
    nllfs_super = val['nllfs_super']
    nllf_diff = (nllfs_pr - nllfs_super) / np.abs(nllfs_super)
    nllf_diff2 = (nllfs_pr - nllfs_super) / np.abs(nllfs_pr)
    nllf_diff[np.isnan(nllf_diff)] = nllf_diff2[np.isnan(nllf_diff)]
    nllf_diff = nllf_diff[np.isfinite(nllf_diff)]
    nllf_diff_good = -nllf_diff[nllf_diff < 0]
    nllf_diff_bad = nllf_diff[nllf_diff > 0]

    ax.hist(np.log10(nllf_diff_good), alpha=0.5)
    ax.hist(np.log10(nllf_diff_bad), alpha=0.5)

    old_nans = np.sum(np.isnan(nllfs_pr) & (~np.isnan(nllfs_super)))
    new_nans = np.sum(np.isnan(nllfs_super) & (~np.isnan(nllfs_pr)))
    same_nans = np.sum(np.isnan(nllfs_super) & np.isnan(nllfs_pr))

    close = np.isclose(nllfs_pr, nllfs_super, rtol=1e-14)
    better = np.sum((nllfs_pr < nllfs_super) & ~close) + old_nans
    worse = np.sum((nllfs_pr > nllfs_super) & ~close) + new_nans
    same = np.sum(close) + same_nans
    assert (better + worse + same) == len(nllfs_super)
    title = title + f"\n{better}+, {worse}-, {same}="
    ax.set_title(title)

plt.xlabel('Histogram of log10 of NLLF relative changes')
plt.ylabel('Frequency')
plt.legend(('PR', 'main'))
fig.suptitle('Histogram of log10 of NLLF relative changes')
plt.show()

```

image
</details>

### CODE REVIEW

**Review by tupui — APPROVED** [review]
LGTM and verified locally, thanks Matt. Waiting for some green CI before merging.

**Comment by mdhaber** [comment]
OK, this was simplified according to discussion in gh-18176.

**Review by tupui — APPROVED** [review]
ok CI is green and this LGTM (also according to Warren's comments on the issue if I got this correctly.) Thanks again Matt

**Pull request merged**
